include/souper/Infer/Interpreter.h

```diff
@@ -80,30 +81,39 @@ struct EvalValue {
     }
   }
 };
+
 using ValueCache = std::unordered_map<souper::Inst *, EvalValue>;
 
-EvalValue evaluateAddNSW(llvm::APInt a, llvm::APInt b);
-EvalValue evaluateAddNUW(llvm::APInt a, llvm::APInt b);
-EvalValue evaluateAddNW(llvm::APInt a, llvm::APInt b);
-EvalValue evaluateSubNSW(llvm::APInt a, llvm::APInt b);
-EvalValue evaluateSubNUW(llvm::APInt a, llvm::APInt b);
-EvalValue evaluateSubNW(llvm::APInt a, llvm::APInt b);
-EvalValue evaluateUDiv(llvm::APInt a, llvm::APInt b);
-EvalValue evaluateURem(llvm::APInt a, llvm::APInt b);
-EvalValue evaluateShl(llvm::APInt a, llvm::APInt b);
-EvalValue evaluateLShr(llvm::APInt a, llvm::APInt b);
-EvalValue evaluateAShr(llvm::APInt a, llvm::APInt b);
-
-EvalValue evaluateInst(Inst* Root, ValueCache &Cache);
-llvm::KnownBits findKnownBits(Inst* I, ValueCache& C, bool PartialEval = true);
-llvm::KnownBits findKnownBitsUsingSolver(Inst *I, Solver *S, std::vector<InstMapping> &PCs);
-llvm::ConstantRange findConstantRange(souper::Inst* I,
-                                      souper::ValueCache& C,
-                                      bool PartialEval = true);
-llvm::ConstantRange findConstantRangeUsingSolver(souper::Inst* I, Solver *S, std::vector<InstMapping> &PCs);
-bool isConcrete(souper::Inst *I, bool ConsiderConsts = true,
-                                 bool ConsiderHoles = true);
-std::string knownBitsString(llvm::KnownBits KB);
+EvalValue evaluateAddNSW(llvm::APInt A, llvm::APInt B);
+EvalValue evaluateAddNUW(llvm::APInt A, llvm::APInt B);
+EvalValue evaluateAddNW(llvm::APInt A, llvm::APInt B);
+EvalValue evaluateSubNSW(llvm::APInt A, llvm::APInt B);
+EvalValue evaluateSubNUW(llvm::APInt A, llvm::APInt B);
+EvalValue evaluateSubNW(llvm::APInt A, llvm::APInt B);
+EvalValue evaluateUDiv(llvm::APInt A, llvm::APInt B);
+EvalValue evaluateURem(llvm::APInt A, llvm::APInt B);
+EvalValue evaluateShl(llvm::APInt A, llvm::APInt B);
+EvalValue evaluateLShr(llvm::APInt A, llvm::APInt B);
+EvalValue evaluateAShr(llvm::APInt A, llvm::APInt B);
+
+  class ConcreteInterpreter {
+    ValueCache _Cache;
+    bool _CacheWritable = false;
```

Comment:
can you explain the purpose of this flag?
also, what is the rationale for prefixing an underscore on these field names?

Reply:
Cache is only writable during the construction of ConcreteInterpreter. See it's ctor how it's being used. We don't want subsequent invocation of evaluateSingleInst to modify the cache. That's why this flag is there.

Rational behind underscore is to distinguish member variables with the local variables in a function body. Imagine a huge function and you are reading it. Some sort of distinction between member variable and local variables helps, I think. I have seen such a usage somewhere at least once in souper code; so I used them here too.

Reply:
Aren't identifiers starting with _ considered reserved for the standard library?
(Not sure about this, the actual rule is probably more complicated.)

Reply:
That maybe the case for class names and other identifiers but this is inside the class.

Reply:
the generally accepted use for underscores is to find names for constructor parameters used only for initialization. don't ever do this for members or for local variables.